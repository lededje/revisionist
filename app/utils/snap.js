import moment from 'moment';

export const snap = (val, click) => Math.round(val / click) * click;

// These time functions actually ceil instead of round.
export const snapTime = (time, click, clickType = 'minutes') => moment(time)
  .add(click - (moment(time).minute() % click), clickType)
  .startOf(clickType);
export const snapDuration = (duration, click, clickType = 'minutes') => duration.add(click - (duration.as(clickType) % click), clickType);
