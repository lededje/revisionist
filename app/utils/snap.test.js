import moment from 'moment';

import { snap, snapTime, snapDuration } from './snap';

describe('snap', () => {
  it.each([[13, 5], [13, 100], [213, 100], [21.123, 10], [-102, 5]])(
    'snaps %i to the nearest %i',
    (val, snapTo) => {
      expect(snap(val, snapTo)).toMatchSnapshot();
    },
  );
});

describe('snapTime', () => {
  it.each([
    ['1907-01-04T08:38:28.000Z', 5, 'minutes'],
    ['1924-11-04T03:06:58.000Z', 1, 'hour'],
    ['1932-11-28T22:52:21.000Z', 10, 'seconds'],
    ['1939-04-24T13:24:29.000Z', 2, 'years'],
    ['1939-11-16T03:20:52.000Z', 10, 'hours'],
    ['2055-06-30T11:44:58.000Z', 7, 'minutes'],
  ])('correctly snaps %s to %i %s increments', (time, snapTo, snapType) => {
    expect(snapTime(time, snapTo, snapType).toISOString()).toMatchSnapshot();
  });
});

describe('snapDuration', () => {
  it.each([
    [11, 'minutes', 5, 'minutes'],
    [21, 'hours', 30, 'hours'],
    [382, 'minutes', 5, 'minutes'],
    [113, 'years', 5, 'year'],
  ])('snaps a moment duration of %i %s to %i %s increments', (amount, type, snapTo, snapType) => {
    expect(
      snapDuration(moment.duration().add(amount, type), snapTo, snapType).as(snapType),
    ).toMatchSnapshot();
  });
});
