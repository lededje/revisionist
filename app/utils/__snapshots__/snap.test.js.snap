// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snap snaps -102 to the nearest 5 1`] = `-100`;

exports[`snap snaps 13 to the nearest 5 1`] = `15`;

exports[`snap snaps 13 to the nearest 100 1`] = `0`;

exports[`snap snaps 21 to the nearest 10 1`] = `20`;

exports[`snap snaps 213 to the nearest 100 1`] = `200`;

exports[`snapDuration snaps a moment duration of 11 minutes to 5 minutes increments 1`] = `15`;

exports[`snapDuration snaps a moment duration of 21 hours to 30 hours increments 1`] = `30`;

exports[`snapDuration snaps a moment duration of 113 years to 5 year increments 1`] = `115`;

exports[`snapDuration snaps a moment duration of 382 minutes to 5 minutes increments 1`] = `385`;

exports[`snapTime correctly snaps 1907-01-04T08:38:28.000Z to 5 minutes increments 1`] = `"1907-01-04T08:40:00.000Z"`;

exports[`snapTime correctly snaps 1924-11-04T03:06:58.000Z to 1 hour increments 1`] = `"1924-11-04T04:00:00.000Z"`;

exports[`snapTime correctly snaps 1932-11-28T22:52:21.000Z to 10 seconds increments 1`] = `"1932-11-28T22:52:29.000Z"`;

exports[`snapTime correctly snaps 1939-04-24T13:24:29.000Z to 2 years increments 1`] = `"1940-12-31T23:00:00.000Z"`;

exports[`snapTime correctly snaps 1939-11-16T03:20:52.000Z to 10 hours increments 1`] = `"1939-11-16T13:00:00.000Z"`;

exports[`snapTime correctly snaps 2055-06-30T11:44:58.000Z to 7 minutes increments 1`] = `"2055-06-30T11:49:00.000Z"`;
